### PR TITLE
[Core] Merge-on-load fast path for diffusion LoRA (shadow wrappers)

### LINKED ISSUE #5053: [Performance]: Diffusion LoRA apply() is memory-bandwidth-bound (+20–30% per forward); proposal: single-adapter merge-on-load fast path
### Proposal to improve performance

**Context.** In RL fine-tuning with verl-omni (BAGEL FlowGRPO, LoRA rank 64), rollout generation is ~28% slower than non-LoRA with identical configs — root-caused to `DiffusionBaseLinearLayerWithLoRA.apply()`. Details and W&B data: verl-project/verl-omni#248.

**Microbenchmark** (isolated 40-layer chain of wrapped `ReplicatedLinear`, dim 3584, bf16, RTX 4090, v0.22.0 — the LoRA layer code is identical on current main; rank 64):

| tokens | wrapper (current) | torch.compile-fused delta | ΔW merged into base |
|--------|-------------------|---------------------------|---------------------|
| 1024   | +12.4%            | +10.4%                    | ≈0%                 |
| 4096   | +19.8%            | +13.0%                    | ≈0%                 |
| 16384  | +29.9%            | +19.7%                    | ≈0%                 |

**Root cause is memory bandwidth, not kernel launches.** The overhead grows with token count and is nearly rank-independent (16/64/128 within 2pp). The two LoRA GEMMs cost only ~12µs/layer; the dominant cost is that `y_flat[slice] = y_flat[slice] + delta` materializes the full-width `(tokens × out_dim)` delta and re-touches the output three extra times per layer (profiler shows an extra Memcpy DtoD per layer from the slice assignment). This bounds what kernel-level fixes can recover: a compiled fused delta gets back only ~1/3, and a punica port (in-place expand) would still leave an estimated ~12% since the shrink/expand passes over `x` and `y` remain.

**Proposal: opt-in single-adapter merge-on-load.** When exactly one adapter is active, fold ΔW = B·A into the base weights at activation time instead of computing the delta every forward:

- In `DiffusionLoRAManager._activate_adapter`: for each target layer, keep a pristine copy of the base weight (GPU by default, optionally CPU-pinned), then write `W = W_pristine + B_eff·A` (scaling is already folded into `lora_b` by `optimize()` + external scale). Leave the wrapper's stacked buffers inactive, so `apply()` takes the existing all-slices-inactive fast return — measured at ~+1% vs bare.
- Deactivate/switch restores from the pristine copy — exact, no accumulation drift, which matters for RL where the adapter is re-synced every training step.
- Guards, falling back to the current path: more than one registered adapter, quantized base layers (can't add ΔW into quantized weights), `fully_sharded_loras`. Under plain TP no communication is needed: `lora_a/b` are already sharded per rank, so ΔW is the local weight shard.
- Cost: one small GEMM + weight-sized write per layer per adapter (de)activation (ms-scale, vs ~70s/step saved in the RL case above); memory for pristine copies of target layers only.

Beyond RL weight syncing, this also benefits the common serving pattern of loading one style LoRA at startup (`lora_path` in `OmniDiffusionConfig`) — same ~20–30% generation speedup at zero per-forward cost.

I have the microbench script (single-GPU, self-contained) and can submit it with the PR. Happy to implement this behind a config flag — feedback on the design welcome, especially on the pristine-copy placement (GPU vs CPU-pinned) and whether auto-enabling for the static single-LoRA path is desirable.

### PR DESCRIPTION
**Pull request by cr-gao**
## Purpose

Fixes #5053. Diffusion LoRA pays two distinct per-forward costs:

- **Eager pipelines** (BAGEL-style `enforce_eager`): `apply()` materializes a full-width delta and re-touches the `(tokens × out_dim)` activation three extra times per layer — memory-bandwidth-bound, +20–30% at realistic token counts (+28% end-to-end in verl-project/verl-omni#248).
- **Compiled pipelines** (SD3.5): installed wrappers fragment the torch.compile'd graph — denoise ticks 647→1549ms (2.4×), CUDA busy 42.6%→22%.

This adds opt-in `lora_merge_on_load` (default off): at activation, ΔW = B·A is folded into base weights and wrappers are kept as *shadow* objects — created for their TP-slicing/packed `set_lora` logic, never installed. The module tree, hence any compiled graph, stays identical to non-LoRA; per-forward cost drops to zero.

Design notes:
- fp32 ΔW accumulation; a pristine copy per touched weight makes deactivate/switch bitwise-exact (no drift across hundreds of RL adapter swaps). Memory cost: ~1× the targeted weights (0.57 GiB for SD3.5-M with 121 rank-32 targets), documented in `docs/user_guide/diffusion/lora.md`.
- TP-safe with no communication: `lora_a/b` are already sharded per rank, so ΔW is the local shard's delta.
- Quantized targets fall back to standard wrapper installation, decided globally at wrap time (shadow and installed modes never mix); an ineligible layer reached at merge time raises rather than silently dropping the adapter.
- Incompatible with `enable_layerwise_offload` (its per-block CPU→GPU weight re-copy would wipe the merged delta): the worker falls back to standard wrappers with a warning. Plain `enable_cpu_offload` is safe (device moves preserve values). Base weights must not be otherwise modified while merged; RL weight-sync flows unmerge via `remove_lora` first.

## Test Plan

**vLLM Version:** v0.22.0 (GPU validation env; `vllm_omni/diffusion/lora` is identical between v0.22.0 and current main)

**vLLM-Omni Commit:** this branch, on latest main. torch 2.11.0+cu130, RTX 4090 24GB.

1. Unit: `pytest tests/diffusion/lora/test_merge_on_load.py -v` — 11 CPU cases: merge math incl. packed/partial slices and scale, bitwise restore, adapter switch without accumulation, merge-time raise on quantized layers, shadow-vs-install wiring incl. creation-time fallback, real-`apply()` before/after output consistency.
2. Microbenchmark: 40-layer wrapped `ReplicatedLinear` chain, dim 3584, bf16, tokens {1024, 4096, 16384} × rank {16, 64, 128}, CUDA-event timing, 10 warmup / 30 measured iters.
3. End-to-end RL: SD3.5-Medium FlowGRPO via verl-omni (`examples/flowgrpo_trainer/sd35/run_sd35_medium_ocr_lora.sh`, 512×512, 10 denoise steps, 64 images/step, rank 32, 121 merged layers, adapter re-synced every training step), three arms in one batch on 3×RTX 4090, 8 steps each; step 1 (warmup + base sync) discarded, reward-server-stall steps identified via the agent-loop timing breakdown and excluded, median of remaining steps.

## Test Result

1. Unit: 11/11 pass.
2. Microbench, rank 64 (overhead vs bare linear; rank 16–128 within 2pp — bandwidth-bound, not compute):

   | tokens | wrapper (current) | ΔW merged |
   |--------|-------------------|-----------|
   | 1024   | +12.4%            | ≈0%       |
   | 4096   | +19.8%            | ≈0%       |
   | 16384  | +29.9%            | ≈0%       |

3. E2E generation/step: no-LoRA 17.1s / wrappers 37.9s (+122%) / merge-on-load **17.9s (+4.7%)**. Rollout-vs-trainer importance ratio `1 ± 1e-6`, same as the wrapper path; merge/unmerge cycles once per adapter update (verified by instrumentation). Peak VRAM not separately recorded; the feature's only steady-state overhead is the pristine copies quantified above.

### CODE REVIEW

**Comment by alex-jw-brooks** [comment]
Thanks @cr-gao. Can you clarify why this would be a needed feature in vLLM Omni, as opposed to just merging the weights into the base model prior to starting the server?

vLLM's lora manager does not support this as a feature since people can just premerge their weights, and it would be ideal to avoid adding special LoRA features for diffusion only IMO

**Comment by cr-gao** [comment]
@alex-jw-brooks Good question! Pre-merging fully covers the static case, but the motivating use case can't do it: at the time of **RL training** (details in verl-project/verl-omni#248): the adapter is re-synced into the *running* server every training step via in-memory `add_lora`. There is no offline moment to pre-merge, and merging trainer-side means shipping the full state dict every step, which is verl-omni's existing `lora.merge=true` path: 10.4s vs 4.2s for LoRA-only sync on BAGEL; worse on larger DiTs. Merge-on-load keeps the rank-sized sync *and* zero-cost inference.
Also, Pre-merging forfeits runtime adapter switching, as N adapters become N baked checkpoints and server restarts; while this path keeps `add_lora`/switch semantics with exact weight restore.
On avoiding diffusion-only LoRA features, I agree as a principle, but the cost asymmetry is itself diffusion-specific. The AR path has punica kernels that make an active adapter cheap, so vLLM never needed merging; the diffusion path has plain-matmul `apply()` (+20–30% eager) and, worse, installed wrappers fragment torch.compile'd pipelines (2.4×). Reaching parity the AR way would mean porting punica *and* making the wrappers compile-transparent, which would be far more invasive than this opt-in flag.
Happy to keep the surface minimal, it's engine-internal today as one config flag; no frontend/protocol changes; default off.

File: vllm_omni/diffusion/lora/manager.py [hunk 1]
```diff
@@ -419,21 +449,50 @@ def _matches_target(module_name: str) -> bool:
...
+                    "installing standard LoRA wrappers instead.",
+                    len(ineligible),
+                    ineligible[0],
                 )
-
-                if lora_layer is not module and isinstance(lora_layer, BaseLayerWithLoRA):
-                    replace_submodule(component, module_name, lora_layer)
-                    self._lora_modules[full_module_name] = lora_layer
-                    logger.debug("Replaced layer: %s -> %s", full_module_name, type(lora_layer).__name__)
+                self.merge_on_load = False
```
Comment (lishunyang12) [inline vllm_omni/diffusion/lora/manager.py]: Mutating `self.merge_on_load` here loses what the user asked for, keeping only the effective mode. Keep the request in a separate `_merge_requested` and flip only an effective flag, so a caller inspecting `manager.merge_on_load` can tell "never requested" from "requested but fell back".

File: vllm_omni/diffusion/lora/manager.py [hunk 2]
```diff
@@ -631,10 +692,106 @@ def _activate_adapter(self, adapter_id: int, scale: float) -> None:
...
+            weight = lora_layer.base_layer.weight
+            if full_module_name not in self._pristine_weights:
+                self._pristine_weights[full_module_name] = weight.detach().clone()
+            weight.data.add_(delta.to(weight.dtype))
+            # Buffers are consumed; reset so apply() takes the inactive fast path.
+            lora_layer.reset_lora(0)
+            merged_layers += 1
+
+        self._merged = merged_layers > 0
+        logger.info("Merged active LoRA into base weights: %d layers", merged_layers)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/lora/manager.py]: This logs at INFO on every activation, a line every step for the RL re-sync-per-step workload this targets (on top of the `Activating adapter` info). Drop it to `logger.debug`.

File: tests/diffusion/lora/test_merge_on_load.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    manager = _make_manager(merge_on_load=True, layer=layer)
+    lora = _make_lora(rank=2, out_dim=6, seed=3)
+    _register(manager, 7, lora)
+
+    w0 = layer.base_layer.weight.detach().clone()
+    manager._activate_adapter(7, scale=1.0)
+    manager._activate_adapter(7, scale=2.0)
+
+    expected = w0 + 2.0 * (lora.lora_b @ lora.lora_a)
+    assert torch.allclose(layer.base_layer.weight, expected, atol=1e-6)
```
Comment (lishunyang12) [inline tests/diffusion/lora/test_merge_on_load.py:171]: None of these cover a rank change while merged, where activating a bigger-rank adapter goes through `_ensure_max_lora_rank` and re-activates the old adapter while `_merged` is still True. Add a case that activates a small-rank then a larger-rank adapter and checks the final merged weight.

**Review by lishunyang12 — COMMENTED** [review]
Nice fast path. The shadow-wrapper and in-place merge design is clean and the bitwise-exact restore is well thought out.

File: vllm_omni/diffusion/lora/manager.py (hunk 1 above)
Comment (cr-gao) [reply vllm_omni/diffusion/lora/manager.py]: Thanks for the suggestions! `merge_on_load` now always holds the request; the effective mode is a separate internal `_merge_enabled`, and only that is cleared on fallback.

File: vllm_omni/diffusion/lora/manager.py (hunk 2 above)
Comment (cr-gao) [reply vllm_omni/diffusion/lora/manager.py]: Done, dropped to `debug`.

File: tests/diffusion/lora/test_merge_on_load.py (hunk 3 above)
Comment (cr-gao) [reply tests/diffusion/lora/test_merge_on_load.py:171]: Added `test_rank_growth_while_merged`, where rank-2 adapter merged while rank-16 adapter forces `_ensure_max_lora_rank`; asserts final weight = pristine + new delta W and bitwise restore on deactivate. Traced the path too, the mid-flight re-activation goes through `_activate_adapter`, which always unmerges first, so it was already correct.

**Comment by cr-gao** [comment]
@alex-jw-brooks Gentle ping — does the RL re-sync rationale above resolve your pre-merge question? All of @lishunyang12's review comments are addressed in 8ed4d082, CI is green.

**Comment by cr-gao** [comment]
Found Intel-ci red, rebased onto main. All concerns are addressed.

**Comment by cr-gao** [comment]
In progress as verl-omni is currently working on weight merge entrypoint which may rely on this. This PR is the baseline of 7557.
